feat: create snapshot delta object with logtsvs

eivindkopperud opened the pull request:
Funksjonalitet for å mate inn LogTSV-er og få ut en liste med snapshots + hele loggen. Den Snapshot-klassen fungerer ikke atm, går kanskje ikke å extende Graph 🤔 ❓

TheodorRene committed:
[WIP] SnapshotDelta implementation

eivindkopperud committed:
merge actions in each interval + restructure

TheodorRene committed:
Add comments,flatMap and fix some possible bugs

See comments in code about the changes in the boolean logic

TheodorRene committed:
Refactor get{Edge,Vertex}Ids into a function

eivindkopperud committed:
apply actions to previous snapshots

eivindkopperud committed:
fix createGraph method

TheodorRene commented on data-generation/src/main/scala/SnapshotDelta.scala:
```diff
@@ -75,55 +75,90 @@ object SnapshotDelta {
+  val edgeIds = logs.flatMap(getEdgeIds)
+
+  // Squash all actions on each edge to one single action
+  val edgesWithAction = edgeIds.map(id => (id, logs.filter(log => log.objectType match {
```
Hmm, håper det er lov å kjøre RDD operasjoner inni en annen RDD operasjon. Burde jo gå ann det

TheodorRene commented on data-generation/src/main/scala/SnapshotDelta.scala:
```diff
@@ -75,55 +75,90 @@ object SnapshotDelta {
+  val joinedEdges = previousSnapshotEdges.fullOuterJoin(edgesWithAction)
+
+  // Create, update or omit (delete) edges based on values are present or not.
+  val newSnapshotEdges = joinedEdges.flatMap(x => x._2 match {
```
Her må man holde tunga rett i munn ja!

TheodorRene replied:
All disse tuple indekseringen irriterer meg, men det er lite å gjøre med

TheodorRene commented on data-generation/src/main/scala/SnapshotDelta.scala:
```diff
@@ -75,55 +75,90 @@ object SnapshotDelta {
+      case newEdge.action == DELETE => None
+      case newEdge.action == UPDATE => Some(Edge(x._1._1, x._1._2, rightWayMergeHashMap(snapshotEdge, newEdge.attributes)))
+    }
+    case (None, Some(newEdge)) => newEdge match {
```
`// The edge has been created in the interval`

TheodorRene replied:
Tror det er fint å få kommentert alle casene. Jeg kan se på det i morgen.

TheodorRene commented on data-generation/src/main/scala/SnapshotDelta.scala:
```diff
@@ -75,55 +75,90 @@ object SnapshotDelta {
+    }
+    case (None, Some(newEdge)) => newEdge match {
+      case newEdge.action == DELETE => None
+      case newEdge.action == UPDATE => Some(Edge(x._1._1, x._1._2, newEdge.attributes))
```
Hva skjer med CREATE? Det er vel likt med UPDATE egentlig.
```
case newEdge.action == CREATE => Some(Edge(x._1._1, x._1._2, newEdge.attributes))
```

TheodorRene commented on data-generation/src/main/scala/SnapshotDelta.scala:
```diff
@@ -75,55 +75,90 @@ object SnapshotDelta {
+    }
+    case (None, Some(newVertex)) => newVertex match {
+      case newVertex.action == DELETE => None
+      case newVertex.action == UPDATE => Some((x._1, newVertex.attributes))
```
Her mangler også. Create. I fix in the morning. Hodet mitt var ikke helt på plass i dag, så fikk ikke gjort så mye

TheodorRene commented on data-generation/src/main/scala/SnapshotDelta.scala:
```diff
@@ -158,6 +145,19 @@ def applyVertexLogs(snapshot: Graph[LTSV.Attributes, LTSV.Attributes], logs: RDD
   newSnapshotVertices
 }
 
+def getSquashedVertexActions(logs: RDD[LogTSV]): RDD[(Long, LogTSV)] = {
```
`groupSquashedVertexActionsById`

TheodorRene replied:
Squashed or Merged. who knows

TheodorRene commented on data-generation/src/main/scala/SnapshotDelta.scala:
```diff
@@ -92,26 +82,17 @@ object SnapshotDelta {
+  }
 }
 def applyEdgeLogs(snapshot: Graph[LTSV.Attributes, LTSV.Attributes], logs: RDD[LogTSV]): RDD[Edge[Attributes]] = {
   val previousSnapshotEdges = snapshot.edges.map(edge => ((edge.dstId, edge.srcId), edge.attr))
```
`previousSnapshotEdgesKV` for å indikere at den er på (Key,Value) format

TheodorRene reviewed (dismissed):
I fikser resten i morgen. Ser sykt bra ut! Bra jobba :D Neste steg etter finpussen er å prøve å kjøre den haha

TheodorRene commented on data-generation/src/main/scala/SnapshotDelta.scala:
```diff
@@ -0,0 +1,234 @@
+  newSnapshotEdges
+}
+
+def getSquashedEdgeActions(logs: RDD[LogTSV]): RDD[((Long, Long), LogTSV)] = {
```
Hmm, her gjør vi antagelsen at det kun kan være 1 kant mellom et nodepar.

TheodorRene commented on data-generation/src/main/scala/SnapshotDelta.scala:
```diff
@@ -0,0 +1,234 @@
+  val joinedEdges = previousSnapshotEdges.fullOuterJoin(squashedEdgeActions)
+
+  // Create, update or omit (delete) edges based on values are present or not.
+  val newSnapshotEdges = joinedEdges.flatMap(x => x._2 match {
```
Når vi squasher så tar vi med en del Vertex Logs også. Disse blir ikke håndtert her. Vi bryr jo oss egentlig kun om DELETES.
Vertex DELETES er det eneste som er relevant for Edges

TheodorRene committed:
Rename variables for clearer intent and fix logical errors

Rename variables and functions, so it's easier to understand what's going on.
Add comments, since many of the cases can be hard to track.
Add CREATE case, since it was missing
Only add Vertex Logs that were DELETE to related edges

TheodorRene commented on data-generation/src/main/scala/SnapshotDelta.scala:
```diff
@@ -87,65 +88,74 @@ object SnapshotDeltaObject {
-      case newEdge.action == UPDATE => Some(Edge(x._1._1, x._1._2, newEdge.attributes))
+    case (None, Some(newEdgeAction)) => newEdgeAction match {                               // Edge was introduced in this interval
+      case newEdgeAction.action == DELETE => None                                           // Edge was introduced then promptly deleted (possibly due to a deleted vertex)
+      case newEdgeAction.action == CREATE =>                                                // Edge was created
```
Denne er ny

TheodorRene commented on data-generation/src/main/scala/SnapshotDelta.scala:
```diff
@@ -87,65 +88,74 @@ object SnapshotDeltaObject {
 }
 
-def getSquashedEdgeActions(logs: RDD[LogTSV]): RDD[((Long, Long), LogTSV)] = {
+// We are makin the assumption that there can only be one edge between two nodes. We might have to create a surrogate key
```
denne kommentaren er viktig

TheodorRene commented on data-generation/src/main/scala/SnapshotDelta.scala:
```diff
@@ -87,65 +88,74 @@ object SnapshotDeltaObject {
-    case VERTEX(objId) => id._1 == objId || id._2 == objId
-    case EDGE(srcId, dstId) => id._1 == srcId && id._2 == dstId
+  val edgesWithAction = edgeIds.map(edgeIDs => (edgeIDs, logs.filter(log => log.objectType match {
+    case VERTEX(objId) => (edgeIDs._1 == objId || edgeIDs._2 == objId) && log.action == DELETE // Vertex DELETES means that the edge is deleted as well
```
dette predikatet er nytt

eivindkopperud replied:
Good catch!

TheodorRene commented on data-generation/src/main/scala/SnapshotDelta.scala:
```diff
@@ -87,65 +88,74 @@ object SnapshotDeltaObject {
-      case newVertex.action == UPDATE => Some((x._1, newVertex.attributes))
+    case (None, Some(newVertex)) => newVertex match {                                 // Vertex was introduced in this interval
+      case newVertex.action == DELETE => None                                         // Vertex was introduced then promptly deleted
+      case newVertex.action == CREATE =>                                              // Vertex was created
```
Denne er ny

TheodorRene committed:
Refactor case expression

TheodorRene commented on data-generation/src/main/scala/SnapshotDelta.scala:
```diff
@@ -216,19 +226,19 @@ def getVertexIds(log: LogTSV): Option[Long] = {
-      case Action.UPDATE => assert (1 == 2); None
+      case Action.CREATE => Some(vertexTuple._1, vertexTuple._2.attributes) // Vertex was created
+      case Action.DELETE => None                                            // Vertex was created and deleted
+      case Action.UPDATE => Some(vertexTuple._1, vertexTuple._2.attributes) // Vertex was created and updated
```
Denne er  ny

TheodorRene commented on data-generation/src/main/scala/SnapshotDelta.scala:
```diff
@@ -216,19 +226,19 @@ def getVertexIds(log: LogTSV): Option[Long] = {
-      case Action.UPDATE => assert (1 == 2); None
+      case Action.CREATE => Some(Edge(edge._1._1, edge._1._2, edge._2.attributes)) // Edge was created
+      case Action.DELETE => None                                                   // Edge was created and deleted (Possibly because of a deleted vertex)
+      case Action.UPDATE => Some(Edge(edge._1._1, edge._1._2, edge._2.attributes)) // Edge was created and updated
```
Denne er ny

eivindkopperud commented on data-generation/src/main/scala/SnapshotDelta.scala:
```diff
@@ -5,6 +5,7 @@ import org.apache.spark.graphx._
 import org.apache.spark.rdd.RDD
 import org.apache.spark.storage.StorageLevel
 
+import scala.collection.mutable
 import scala.collection.mutable.MutableList
```
Funker det ikke å importere MutableList direkte?

TheodorRene replied:
hmm, kan se på det. Det var intellij om klagde. Jeg bare gjorde som den sa. Den satt det som en warning

eivindkopperud commented on data-generation/src/main/scala/SnapshotDelta.scala:
```diff
@@ -61,13 +62,13 @@ object SnapshotDeltaObject {
   }
 
-  def snapShotCount(logs: RDD[LogTSV], numberOfActions: Int): SnapshotDelta[Attributes, Attributes] = {
+  def createSnapShotCountModel(logs: RDD[LogTSV], numberOfActions: Int): SnapshotDelta[Attributes, Attributes] = {
```
```suggestion
  def createSnapshotCountModel(logs: RDD[LogTSV], numberOfActions: Int): SnapshotDelta[Attributes, Attributes] = {
```
SnapShot -> Snapshot

eivindkopperud commented on data-generation/src/main/scala/SnapshotDelta.scala:
```diff
@@ -87,65 +88,74 @@ object SnapshotDeltaObject {
+      case DELETE => None                                           // Edge was introduced then promptly deleted (possibly due to a deleted vertex)
+      case CREATE =>                                                // Edge was created
+        Some(Edge(outerJoinedEdge._1._1, outerJoinedEdge._1._2, newEdgeAction.attributes))
+      case UPDATE =>                                                // Edge was created and had an update
+        Some(Edge(outerJoinedEdge._1._1, outerJoinedEdge._1._2, newEdgeAction.attributes))
```
Dette skal vel kanskje ikke skje? Hvis den er ny dette intervallet, kan den enten være CREATE eller DELETE?

eivindkopperud commented on data-generation/src/main/scala/SnapshotDelta.scala:
```diff
@@ -87,65 +88,74 @@ object SnapshotDeltaObject {
+      case DELETE => None                                                             // Vertex was introduced then promptly deleted
+      case CREATE =>                                                                  // Vertex was created
+        Some((outerJoinedVertex._1, newVertex.attributes))
+      case UPDATE =>                                                                  // Vertex was created and had an update
+        Some((outerJoinedVertex._1, newVertex.attributes))
```
Dette skal vel heller ikke skje

TheodorRene committed:
Update data-generation/src/main/scala/SnapshotDelta.scala

Co-authored-by: Eivind Kopperud <eivind.kop@gmail.com>

TheodorRene committed:
Revert added UPDATE case for when entities are new

CREATE,UPDATE gets merged into a CREATE in an interval

TheodorRene committed:
Fix renaming of function

eivindkopperud reviewed (commented):
**approves in comment since github doesnt allow authors to approve their own pr**
Ser bra ut nå synes jeg!
